[Woo POS] Hide cart images when the view is too narrow

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->


<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->
Closes: #15053

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR adds a minimum width criteria for showing the images on CartItems, as they are often shown in a fairly constrained width, especially in portrait or landscape 2/3rds views.

1/2 landscape view on a 13" iPad never shows images – it's pretty rough even at the smallest sizes if we show them, at least when there's a remove button shown as well. On the Checkout screen it looks like there's space, but it's strange for the cart items to change as you move between those two screens.

The dynamicTypeSize and width go together to decide whether the images should be shown, so the minimum width changes depending on your current type size.

Note that I chose to do the calculations on the cart, not each image. It just felt bad for performance to have lots of views using geometry readers to calculate the exact same answer... but it would be nicer from a code-readability perspective to have it on the item cards.

Another nice step would be to update `DynamicHStack` to respond to width constraints as well – at the moment, moving to an accessibilty size avoids the `Cart` title having to wrap, but we could do that at smaller sizes when in a narrower display mode. I think that should be done separately though.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Launch the app and navigate to POS
2. Add an item to the cart
3. Observe that at 100% text size (and some sizes above) in landscape fullscreen, the cart image is shown.
4. Navigate to the checkout
5. Observe that the cart image is still shown (or hidden) as it was on the previous screen.

Explore different screen size splits, orientations, and dynamic type sizes.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I've tested this by moving between the item list and checkout, in portrait, landscape 2/3rds, and landscape views.

I used an iPad Air on 17.5, an iPad mini simulator, and an iPad Air 13" simulator.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/60cc97c6-0b96-4dc1-9990-e491ead1a128


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

joshheald committed:
Only show cart item images when it’s wide enough

joshheald committed:
Cell padding when the cart item image is hidden

joshheald committed:
Show leading padding on cart items when no image

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
Rely on HStack spacing between x button and text

joshheald committed:
Make image hiding more reliable by passing width

joshheald committed:
Remove development log

staskus commented on WooCommerce/Classes/POS/Presentation/CartView.swift:
```diff
@@ -156,6 +168,39 @@ private extension CartView {
+        shouldShowItemImages = newVisibility
+    }
+
+    func minimumWidthToShowItemImages(with dynamicTypeSize: DynamicTypeSize) -> CGFloat {
```
> Note that I chose to do the calculations on the cart, not each image. It just felt bad for performance to have lots of views using geometry readers to calculate the exact same answer... but it would be nicer from a code-readability perspective to have it on the item cards.

Theoretically, we could pass values of the geometry reader to the item view, and then it could decide whether it's shown or not. Of course, this way, responsibilities would be split between two views.

joshheald replied:
Yeah, if we ever want it to be different between different types of cart item then we could do that, but it doesn't seem worth it just now

staskus reviewed (approved):
Thanks for the review!

> I also noticed another unrelated issue with Cart while testing:
> - https://github.com/woocommerce/woocommerce-ios/issues/15053

Fixed in 5119757

joshheald committed:
Avoid secondary background changing in split view

We previously used `tertiarySystemBackground` as our dark mode colour for `posSecondaryBackground`.

This split view support work has shown that in dark mode, Apple’s tertiary background actually changes colour when in split view.

Since we don’t use system colours for our other backgrounds, this can result in views becoming indistinguishable from one another.

This commit defines the dark mode secondary background to match the system background in non-splitview mode.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/237">21.7</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
Merge branch 'trunk' into task/hide-cart-images-when-the-view-is-too-narrow

joshheald committed:
Merge branch 'trunk' into task/hide-cart-images-when-the-view-is-too-narrow
